bullet/src/KinematicsFeatures.cc

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (C) 2020 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include <ignition/common/Console.hh>
+#include "KinematicsFeatures.hh"
+
+namespace ignition {
+namespace physics {
+namespace bullet {
+
+/////////////////////////////////////////////////
+FrameData3d KinematicsFeatures::FrameDataRelativeToWorld(
+    const FrameID &_id) const
+{
+  FrameData3d data;
+
+  // The feature system should never send us the world ID.
+  if (_id.IsWorld())
+  {
+    ignerr << "Given a FrameID belonging to the world. This should not be "
+           << "possible! Please report this bug!\n";
+    assert(false);
+    return data;
+  }
+
+  const auto linkID = _id.ID();
+
+  if (this->links.find(linkID) == this->links.end())
+  {
+    ignerr << "Given a FrameID not belonging to a link.\n";
+    return data;
+  }
+  const auto &linkInfo = this->links.at(linkID);
+  const auto &model = linkInfo->link;
+
+  btTransform trans;
+  model->getMotionState()->getWorldTransform(trans);
+  const btVector3 pos = trans.getOrigin();
+  const btMatrix3x3 mat = trans.getBasis();
+
+  auto eigenMat = convert(mat);
+  auto eigenVec = convert(pos);
+
+  data.pose.linear() = eigenMat;
+  data.pose.translation() = eigenVec;
+
+  // Add base velocities
+  btVector3 omega = model->getAngularVelocity();
+  btVector3 vel = model->getLinearVelocity();
+
+  // Transform to world frame
+  // const auto matBaseToWorld =
+  //   btMatrix3x3(model->getWorldToBaseRot()).inverse();
+  // omega = matBaseToWorld * omega;
+  // vel = matBaseToWorld * vel;
+
+  data.linearVelocity = convert(vel);
```

Comment:
Please check dffee9f to see if I understood you correctly

Reply:
Looking at [code](https://github.com/osrf/gazebo/blob/ea72e4d624ffef409694c6624fffd43b6a946667/gazebo/physics/bullet/BulletLink.cc#L415-L434) in Gazebo classic, I think we need to rotate `inertialPose.Pos()` into the world frame because both `vel` and `omega` are given in the world frame at this point.

The code there actually uses[ `btRigidBody::getVelocityInLocalPoint`](https://github.com/bulletphysics/bullet3/blob/master/src/BulletDynamics/Dynamics/btRigidBody.h#L460-L463), but that is simply doing the cross product as well.

Reply:
I dont have a proper way to test this, but 024bdb2 should take care, at least the math seems to check

Reply:
This changed the order of operands in the cross product. I'm not sure that's correct. The formula is usually given as `omega x p` where `p` is the vector from the COM to the point of interest. But in this case `inertiaPose.Pos()` is the vector from the link origin to COM, so we should either reverse the order of operands or add a `-` sign to `inertiaPose.Pos()`.

Reply:
040bc9e should fix this, thank you @azeey for the help